src/extension/messaging.ts

```diff
@@ -0,0 +1,120 @@
+import { join } from 'node:path'
+
+import { Disposable, ExtensionContext, Uri, window, workspace } from 'vscode'
+import { TelemetryReporter } from 'vscode-telemetry'
+
+import { EXTENSION_NAME, TELEMETRY_EVENTS } from '../constants'
+
+import { fileOrDirectoryExists, getDefaultWorkspace } from './utils'
+
+export abstract class DisplayableMessage {
+    abstract dispose(): void
+    abstract display(): void
+    constructor(protected readonly context: ExtensionContext, readonly stateSettings?: Record<TELEMETRY_EVENTS, any>) {
+        this.context = context
+        this.stateSettings = stateSettings
+    }
+}
+
+export class MessagingBuilder implements Disposable {
+    /**
+     * Represents a collection of DisplayableMessages, useful to display window messages.
+     * @param messages Specify the displayable messages to show
+     */
+    constructor(private messages: DisplayableMessage[]) {
+        this.messages = messages
+    }
+
+    dispose() {
+        this.messages.forEach(d => d.dispose())
+    }
+
+    /**
+     * Display registered DisplayableMessages
+     */
+    activate() {
+        for (const message of this.messages) {
+            message.display()
+        }
+    }
+}
+
+
+export class RecommendExtensionMessage extends DisplayableMessage implements Disposable {
+    async display(): Promise<void> {
+        try {
```

Comment:
I wonder if it makes sense to wrap this all into a try/catch block. It could potentially swallow errors that we don't expect.